xgcm/grid_ufunc.py

```diff
@@ -104,23 +100,84 @@ def _parse_grid_ufunc_signature(
     return in_ax_names, out_ax_names, in_ax_pos, out_ax_pos
 
 
+def _build_signature_from_type_hints(hints: Dict[str, Any]) -> str:
+
+    if "return" in hints:
+        return_hint = hints.pop("return")
+
+        # if ufunc returns multiple values (each of which might be annotated) we must extract from Tuple first
+        return_hints = (
+            list(return_hint.__args__)
+            if return_hint._name == "Tuple"
+            else [return_hint]
+        )
+
+        return_annotations = [
+            hint.__metadata__[0]
+            for hint in return_hints
+            if hasattr(hint, "__metadata__")
+        ]
+
+        return_arg_signature = ",".join(
+            [f"({variable_annotation})" for variable_annotation in return_annotations]
+        )
+    else:
+        return_arg_signature = "()"
+
+    arg_annotations = {
+        argname: hint.__metadata__[0]
+        for argname, hint in hints.items()
+        if hasattr(hint, "__metadata__")
+    }
+
+    # TODO check number of annotations?
+
+    if len(arg_annotations) > 0:
+        input_arg_signature = ",".join(
+            [
+                f"({variable_annotation})"
+                for variable_annotation in arg_annotations.values()
+            ]
+        )
+    else:
+        input_arg_signature = "()"
+
+    return f"{input_arg_signature}->{return_arg_signature}"
```

Comment:
Rather than parsing the type hints to a str input signature, could not not go directly to a `Signature` object? Seems a little inefficient to use the signature formatted string as our internal serialization format of the signature.

Reply:
Yep can do - the design as it is might be slightly weird just because I actually wrote this PR before I implemented the `Signature` class in #441 . You're right that it makes sense to build strings from the internal representation though - I'll change it.

Reply:
I understand what you mean, and I agree. I was referring to (really thinking aloud to myself) about how if I do that then the `__str__` method of `Signature` should rebuild the string from the internal representation. Let me just have a go at it and I think you'll see we are on the same page.

Reply:
Okay I've rewritten it so the constructors follow the pattern you specified @rabernat